nanogateway.py: use ticks_add() and ticks_diff() instead + and -

robert-hh opened the pull request:
The time stamps returned by ticks_ms() overflow after a while.
Therefore they should not simply be added or subtracted.
ticks_add(9 and ticks_diff() are to be used instead.

Also: pull upfront the downlink send time by 8 ms to cope with internal
delays. Then there is a better match with the receive window1.

robert-hh committed:
nanogateway.py: use ticks_add() and ticks_diff() instead + and -

The time stamps returned by ticks_ms() overflow after a while.
Therefore they should not simply be added or subtracted.
ticks_add(9 and ticks_diff() are to be used instead.

Also: pull upfront the downlink send time by 8 ms to cope with internal
delays. Then there is a better match with the receive window1.

danicampora commented on examples/lorawan-nano-gateway/nanogateway.py:
```diff
@@ -369,16 +369,14 @@ def _udp_thread(self):
-                                tx_pk["txpk"]["tmst"] - 50, tx_pk["txpk"]["datr"],
-                                int(tx_pk["txpk"]["freq"] * 1000) * 1000
+                                tmst, tx_pk["txpk"]["datr"],
+                                int(tx_pk["txpk"]["freq"] * 1000 + 0.0005) * 1000
```
@robert-hh thanks for this contribution. One question:

Why the `` + 0.0005`` here?

robert-hh commented:
In order to round the frequency, which is supplied as float, to the nearest integer. Without that, the frequency was set to 868099000 Hz. Technically that might not matter, but it did not look nice.

robert-hh committed:
nanogateway.py: Reduced the timing compensation to 4 ms

The gateway looses at least about 4 ms when handling the receive window 1,
consisting of:

time end of transmission node -> taking the time stamp gateway: 0.7 ms
time gateway calling send -> start of RF transmission: 3.2 ms

robert-hh committed:
nanogateway: Added sleep_ms(1) in the waiting loop for downlink

So it#s a little bit les busy waiting

robert-hh committed:
nanogateway: Set window1 timing compensation to 10ms. Adapt for 1.17.3.b1

The nanogateway tends to be 10ms late, reasononly partially known (for 4 ms).
This change compensates for that.
Chnage the call from ticks_us() to ticks_cpu(), as result from
changes in 1.17.3.b1

robert-hh committed:
nanogateway: correct send time based on the mesage length

robert-hh committed:
nanogateway.py: set sockettimeout before lora send

The results in the time between the send call and the actual RF
start being practially independent form the message  length. The
oberved delay is between 600 and 1600 µs. For that, the send time
is puule dupfront by 800µs.

robert-hh committed:
nanogateweay: Adapt timing correction based on the board type

For LoPy 1, it's 1ms for LoPy4 and FiPy it's 6 ms.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pycom/pycom-libraries?pullRequest=54) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pycom/pycom-libraries?pullRequest=54) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pycom/pycom-libraries?pullRequest=54) it.</sub>
